### PR TITLE
[do not merge!] Read region implementation

### PR DESCRIPTION
**Pull request by alex-virodov**
This is a request for comments - those are the changes backported from my OpenSlide integration. Please take a look when you have a chance. 

This is very far from the nice api we discussed, but is actually sufficient to proceed with OpenSlide integration I think. The huge benefit is that we will no longer have 2 copies of the code, even with this far-from-perfect api.

The `isyntax_example.c` is working for me and produces a correctly-looking png for a test slide I have. (inter-level pixel offsets not corrected here).

I'm pretty sure I broke `isyntax_streamer.c`, should be easy to fix api-wise, but I may have introduced bugs. I probably need a test program for streamer, do you know if easy to write?

In any case, my next task is to use this version of the code with OpenSlide. If it works, we can then pass this repo as-is to @bgilbert for build system integration, as we continue working on the implementation code.

Thoughts and comments are welcomed. I will also leave some comments in the code below after I create the PR.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.24)
+cmake_minimum_required(VERSION 3.23)
```
Comment (alex-virodov) [inline CMakeLists.txt:1]: Clion built-in cmake has this version, so if you don't care, I prefer that. Otherwise happy to bump to what's convenient for you and fix my cmake setup.

File: src/isyntax/isyntax.c [hunk 2]
```diff
@@ -1774,14 +1774,13 @@ u32 isyntax_idwt_tile_for_color_channel(isyntax_t* isyntax, isyntax_image_t* wsi
 	return invalid_edges;
 }
 
-u32* isyntax_load_tile(isyntax_t* isyntax, isyntax_image_t* wsi, i32 scale, i32 tile_x, i32 tile_y) {
+u32* isyntax_load_tile(isyntax_t* isyntax, isyntax_image_t* wsi, i32 scale, i32 tile_x, i32 tile_y, block_allocator_t* ll_coeff_block_allocator, bool decode_rgb) {
```
Comment (alex-virodov) [inline src/isyntax/isyntax.c:1777]: I need explicit block allocator passing because block allocators can now be shared between isyntax_t instances. This is to allow eviction of "another isyntax's tiles".

decode_rgb flag is needed to avoid decoding rgb for intermediate tiles. I know slidescape uses and caches those, so you may be calling this with decode_rgb=true always.

File: src/isyntax/isyntax.c [hunk 3]
```diff
@@ -1817,95 +1816,102 @@ u32* isyntax_load_tile(isyntax_t* isyntax, isyntax_image_t* wsi, i32 scale, i32
...
 
-		// Distribute result to child tiles
-		if (scale > 0) {
-			isyntax_level_t* next_level = wsi->levels + (scale - 1);
-			isyntax_tile_t* child_top_left = next_level->tiles + (tile_y*2) * next_level->width_in_tiles + (tile_x*2);
-			isyntax_tile_t* child_top_right = child_top_left + 1;
-			isyntax_tile_t* child_bottom_left = child_top_left + next_level->width_in_tiles;
-			isyntax_tile_t* child_bottom_right = child_bottom_left + 1;
-
-			ASSERT(child_top_left->color_channels[color].coeff_ll == NULL);
```
Comment (alex-virodov) [inline src/isyntax/isyntax.c:1828]: Note actual logic change below, as I now allow tiles to have coefficients that may need to be released. This is because some tiles may be evicted and some not, so I can't expect them all to be either null or non-null.

File: src/isyntax/isyntax.c [hunk 4]
```diff
@@ -1817,95 +1816,102 @@ u32* isyntax_load_tile(isyntax_t* isyntax, isyntax_image_t* wsi, i32 scale, i32
...
+			if (invalid_edges != 0) {
+				console_print_error("load: scale=%d x=%d y=%d  idwt time =%g  invalid edges=%x\n", scale, tile_x, tile_y, elapsed_idwt, invalid_edges);
+				// early out
+				release_temp_memory(&temp_memory);
+				return NULL;
 			}
 		}
 	}
 
-	tile->is_loaded = true; // Meaning: it is now safe to start loading 'child' tiles of the next level
```
Comment (alex-virodov) [inline src/isyntax/isyntax.c:1908]: this needs to come back (easier to see those on a PR).

File: src/isyntax/isyntax.c [hunk 5]
```diff
@@ -1919,10 +1925,9 @@ u32* isyntax_load_tile(isyntax_t* isyntax, isyntax_image_t* wsi, i32 scale, i32
 	i32 valid_offset = (first_valid_pixel * idwt_stride) + first_valid_pixel;
 	u32* bgra = convert_ycocg_to_bgra_block(Y + valid_offset, Co + valid_offset, Cg + valid_offset,
 											tile_width, tile_height, idwt_stride);
-	isyntax->total_rgb_transform_time += get_seconds_elapsed(start, get_clock());
```
Comment (alex-virodov) [inline src/isyntax/isyntax.c:1922]: this and the timing next few lines need to be reverted, unrelated debug changes.

File: src/isyntax/isyntax.c [hunk 6]
```diff
@@ -2606,7 +2611,7 @@ void isyntax_set_work_queue(isyntax_t* isyntax, work_queue_t* work_queue) {
 	isyntax->work_submission_queue = work_queue;
 }
 
-bool isyntax_open(isyntax_t* isyntax, const char* filename) {
+bool isyntax_open(isyntax_t* isyntax, const char* filename, bool init_allocators) {
```
Comment (alex-virodov) [inline src/isyntax/isyntax.c:2614]: note the api change - I want to be able to suppress default per-isyntax allocators, as they now can be shared. Not perfect API, but works.

File: src/isyntax/isyntax.c [hunk 7]
```diff
@@ -3068,12 +3073,35 @@ bool isyntax_open(isyntax_t* isyntax, const char* filename) {
...
+				isyntax->ll_coeff_block_allocator = NULL;
+				isyntax->h_coeff_block_allocator = NULL;
+				isyntax->is_block_allocator_owned = false;
+			}
 
 			success = true;
 
 			free(read_buffer);
+
+			// Populate debug info.
```
Comment (alex-virodov) [inline src/isyntax/isyntax.c:3093]: 'debug' is a bad name for it, I now rely on these 3 fields for more than debug. Need to either make them proper fields, or move to some external struct. Not sure which is best.

File: src/isyntax/isyntax_reader.c
Comment (alex-virodov) [inline src/isyntax/isyntax_reader.c:1]: you don't have to review this one, but conceptually it is similar to the `isyntax_streamer.c`, in that it figures out the list of tiles to decode in the proper order, and does the cache management in the process. my `isyntax_cache_t` is somewhat similar in spirit to your `isyntax_streamer_t` - in the sense that it holds state.

File: CMakeLists.txt (hunk 1 above)
Comment (Falcury) [reply CMakeLists.txt:1]: I suspect my CLion's CMake was a bit newer when I created the file. We are not using any fancy CMake features though, we can probably also drop it down much lower.

File: src/isyntax/isyntax.c (hunk 2 above)
Comment (Falcury) [reply src/isyntax/isyntax.c:1777]: OK!
One question regarding the sharing of block allocators for multiple isytaxes: will this still work if the codeblock size is different? (Maybe not so likely to happen, but I think the spec allows for this)

File: src/isyntax/isyntax.c [hunk 8]
```diff
@@ -1817,95 +1816,102 @@ u32* isyntax_load_tile(isyntax_t* isyntax, isyntax_image_t* wsi, i32 scale, i32
...
 
-			// After the last color channel, we can report that the children now have their LL blocks available.
-			if (color == 2) {
-				child_top_left->has_ll = true;
-				child_top_right->has_ll = true;
-				child_bottom_left->has_ll = true;
-				child_bottom_right->has_ll = true;
-
-				// Even if the parent tile has invalid edges around the outside, its child LL blocks will still have valid edges on the inside.
-				child_top_left->ll_invalid_edges = invalid_edges & ~(ISYNTAX_ADJ_TILE_CENTER_RIGHT | ISYNTAX_ADJ_TILE_BOTTOM_RIGHT | ISYNTAX_ADJ_TILE_BOTTOM_CENTER);
```
Comment (Falcury) [inline src/isyntax/isyntax.c:1892]: After deleting this the `isyntax_tile_t.ll_invalid_edges` field has no uses anymore. Can it be deleted?

File: src/isyntax/isyntax.c (hunk 3 above)
Comment (Falcury) [reply src/isyntax/isyntax.c:1828]: OK

File: src/isyntax/isyntax.c (hunk 4 above)
Comment (Falcury) [reply src/isyntax/isyntax.c:1908]: Note that `is_loaded`, `is_submitted_for_loading`, `is_submitted_for_h_coeff_decompression` are used by isyntax_streamer.c

File: src/isyntax/isyntax.c (hunk 5 above)
Comment (Falcury) [reply src/isyntax/isyntax.c:1922]: What do we want to do with the benchmarking instrumentation code? Leave it in?

File: src/isyntax/isyntax.c (hunk 6 above)
Comment (Falcury) [reply src/isyntax/isyntax.c:2614]: Makes sense. Maybe we can refine it later.

File: src/isyntax/isyntax.c (hunk 7 above)
Comment (Falcury) [reply src/isyntax/isyntax.c:3093]: I think I didn't need them in isyntax_streamer.c because I always iterated explicitly using arrays and got the indices from there. I see why they're handy for linked lists, though. As long as we're actively using them I guess it's best for them to be proper fields.

Some thoughts going around in my head:
- Iterating over tiles linked lists instead of arrays: is the performance impact minimal enough that we don't need to worry about that?
- What is the best solution for cache management, and how does it link back to this?
- Iterating a region in terms of a list of tiles maybe obfuscates necessary geometric organization (i.e. edges, centers, neighbors etc.) that may be needed for efficient prioritizing (e.g. loading the tile in the center of the region/screen first) or efficient ordering of operations? (I tried to do such things as optimally as I could think of at the time for isyntax_streamer.c)

File: src/isyntax/isyntax.c (hunk 2 above)
Comment (alex-virodov) [reply src/isyntax/isyntax.c:1777]: > will this still work if the codeblock size is different

Not as-is, I have a todo in OpenSlide code, but I currently assume all iSyntax use same block size. I will add some asserts to make sure the shared block allocator is of the right size at runtime.

File: src/isyntax/isyntax.c (hunk 5 above)
Comment (alex-virodov) [reply src/isyntax/isyntax.c:1922]: for now, yes. I don't see a reason to remove it really, and it can be useful.

File: src/isyntax/isyntax.c (hunk 7 above)
Comment (alex-virodov) [reply src/isyntax/isyntax.c:3093]: Good thoughts, thank you. (That's why I did this PR).

> ... linked lists instead of arrays: is the performance impact minimal enough ... ?

I did not measure performance impact here. I chose linked list because moving elements around is O(1). The cache lists are effectively priority queues, maybe heap would be even more appropriate. But I want to leave such optimizations for later. In my latest measurements iirc, hulsken_decompress was taking >50% of runtime code - that is, doing actual useful work.

> What is the best solution for cache management

I implemented LRU cache (accounting for data dependencies), at least that was my intent modulo bugs. This should work for viewing too if you want to evict. If you don't care about releasing memory (viewing one slide at a time, this may be fine), then you don't really need an LRU cache. I think cache is orthogonal to request sequence, and the _streamer code implements mostly sequence, and _reader code implements mostly cache.

> Iterating a region ..  obfuscates .. geometric organization ... for efficient prioritizing 

I think you are assuming an access pattern, which is good when you can do that :) In ML loads for example, I have sequential access and random access both common, and those patterns won't benefit from such optimizations. But they require other optimizations - namely, cache eviction, to keep memory usage in check. As I don't really know the access pattern, LRU is the best I can do. I did an example, it should work well for linear access at the least.

Within the constructed dependency list, we can do additional optimizations, such as extra reads if beneficial, or reordering. But my assumed access pattern is "one tile at a time". Even if we were to support regions, that would bypass OpenSlide tile caching - or I will have to do complex compute of "what region corresponds to a tile". This is why I prefer tile-level access in libisyntax.

Another downside of exposing read_region as opposed to read_tile from libisyntax is that I will have to do sub-tile pixel management - cropping, concatenating, and the memory management needed for that. Tiled formats are common, and OpenSlide already handles all of that, and I assume other libraries/users would too. I agree it disables some optimizations (e.g. suppressing unnecessary neighbor decoding), but that's a tradeoff of general vs specialized. Later this is something we can control with extra flags if it is really needed.

**Comment by alex-virodov** [comment]
I will close this PR, and will make a few smaller ones to get to same state. Thank you for the discussion, please feel free to continue where relevant.

**Comment by Falcury** [comment]
> I'm pretty sure I broke `isyntax_streamer.c`, should be easy to fix api-wise, but I may have introduced bugs. I probably need a test program for streamer, do you know if easy to write?

I can give this a shot. I broke it myself too, it's still too intertwined with the Slidescape codebase to use on its own (I'll try to fix this).
